<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
      <title> | Apizr </title>
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <meta name="title" content=" | Apizr ">
      
      
      <link rel="icon" href="../resources/favicon.ico">
      <link rel="stylesheet" href="../public/docfx.min.css">
      <link rel="stylesheet" href="../public/main.css">
      <meta name="docfx:navrel" content="../toc.html">
      <meta name="docfx:tocrel" content="toc.html">
      
      <meta name="docfx:rel" content="../">
      
      
      <meta name="docfx:docurl" content="https://github.com/Respawnsive/Apizr/blob/dev/Apizr/Docs/Apizr.Docs/articles/config_exceptions.md/#L1">
      <meta name="loc:inThisArticle" content="In this article">
      <meta name="loc:searchResultsCount" content="{count} results for &quot;{query}&quot;">
      <meta name="loc:searchNoResults" content="No results for &quot;{query}&quot;">
      <meta name="loc:tocFilter" content="Filter by title">
      <meta name="loc:nextArticle" content="Next">
      <meta name="loc:prevArticle" content="Previous">
      <meta name="loc:themeLight" content="Light">
      <meta name="loc:themeDark" content="Dark">
      <meta name="loc:themeAuto" content="Auto">
      <meta name="loc:changeTheme" content="Change theme">
      <meta name="loc:copy" content="Copy">
      <meta name="loc:downloadPdf" content="Download PDF">

      <script type="module" src="./../public/docfx.min.js"></script>

      <script>
        const theme = localStorage.getItem('theme') || 'auto'
        document.documentElement.setAttribute('data-bs-theme', theme === 'auto' ? (window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light') : theme)
      </script>

  </head>

  <body class="tex2jax_ignore" data-layout="" data-yaml-mime="">
    <header class="bg-body border-bottom">
      <nav id="autocollapse" class="navbar navbar-expand-md" role="navigation">
        <div class="container-xxl flex-nowrap">
          <a class="navbar-brand" href="../index.html">
            <img id="logo" class="svg" src="../resources/logo.png" alt="">
            
          </a>
          <button class="btn btn-lg d-md-none border-0" type="button" data-bs-toggle="collapse" data-bs-target="#navpanel" aria-controls="navpanel" aria-expanded="false" aria-label="Toggle navigation">
            <i class="bi bi-three-dots"></i>
          </button>
          <div class="collapse navbar-collapse" id="navpanel">
            <div id="navbar">
              <form class="search" role="search" id="search">
                <i class="bi bi-search"></i>
                <input class="form-control" id="search-query" type="search" disabled="" placeholder="Search" autocomplete="off" aria-label="Search">
              </form>
            </div>
          </div>
        </div>
      </nav>
    </header>

    <main class="container-xxl">
      <div class="toc-offcanvas">
        <div class="offcanvas-md offcanvas-start" tabindex="-1" id="tocOffcanvas" aria-labelledby="tocOffcanvasLabel">
          <div class="offcanvas-header">
            <h5 class="offcanvas-title" id="tocOffcanvasLabel">Table of Contents</h5>
            <button type="button" class="btn-close" data-bs-dismiss="offcanvas" data-bs-target="#tocOffcanvas" aria-label="Close"></button>
          </div>
          <div class="offcanvas-body">
            <nav class="toc" id="toc"></nav>
          </div>
        </div>
      </div>

      <div class="content">
        <div class="actionbar">
          <button class="btn btn-lg border-0 d-md-none" type="button" data-bs-toggle="offcanvas" data-bs-target="#tocOffcanvas" aria-controls="tocOffcanvas" aria-expanded="false" aria-label="Show table of contents">
            <i class="bi bi-list"></i>
          </button>

          <nav id="breadcrumb"></nav>
        </div>

        <article data-uid="">

<h2 id="configuring-exception-handling">Configuring Exception handling</h2>
<h3 id="using-trycatch">Using <code>Try/Catch</code></h3>
<p>Here is an example of how to send a request.</p>
<p>Inject <code>IApizrManager&lt;IYourDefinedInterface&gt;</code> where you need it</p>
<pre><code class="lang-csharp">IList&lt;User&gt;? users;
try
{
    var userList = await _reqResManager.ExecuteAsync(api =&gt; api.GetUsersAsync());
    users = userList.Data;
}
catch (ApizrException&lt;UserList&gt; e)
{
    var message = e.InnerException is IOException ? &quot;No network&quot; : (e.Message ?? &quot;Error&quot;);
    Alert.Show(&quot;Error&quot;, message);

    users = e.CachedResult?.Data;
}

if(users != null)
    Users = new ObservableCollection&lt;User&gt;(users);
</code></pre>
<p>We catch any ApizrException as it will contain the original inner exception, but also the previously cached result if some.
If you provided an IConnectivityHandler implementation and there's no network connectivity before sending request, Apizr will throw an IO inner exception without sending the request.</p>
<h3 id="using-iapizrresponse-or-iapizrresponset">Using <code>IApizrResponse</code> or <code>IApizrResponse&lt;T&gt;</code></h3>
<p>Refit has different exception handling behavior depending on if your Refit interface methods return <code>Task&lt;T&gt;</code> or if they return <code>Task&lt;IApiResponse&gt;</code>, <code>Task&lt;IApiResponse&lt;T&gt;&gt;</code>, or <code>Task&lt;ApiResponse&lt;T&gt;&gt;</code>.</p>
<p>When returning <code>Task&lt;IApiResponse&gt;</code>, <code>Task&lt;IApiResponse&lt;T&gt;&gt;</code>, or <code>Task&lt;ApiResponse&lt;T&gt;&gt;</code> <strong>(not <code>Apizr</code> but <code>Api</code>)</strong>,
Refit traps any <code>ApiException</code> raised by the <code>ExceptionFactory</code> when processing the response, and any errors that occur when attempting to deserialize the response to <code>ApiResponse&lt;T&gt;</code>, and populates the exception into the <code>Error</code> property on <code>ApiResponse&lt;T&gt;</code> without throwing the exception.</p>
<p>Then, Apizr will wrap the <code>ApiResponse&lt;T&gt;</code> into an <code>ApizrResponse&lt;T&gt;</code> plus some cached data if any and return it as a final response.
You can then decide what to do like so:</p>
<pre><code class="lang-csharp">// Here we wrap the response into an IApiResponse&lt;T&gt; provided by Refit
[BaseAddress(&quot;https://reqres.in/api&quot;)]
public interface IReqResService
{
    [Get(&quot;/users&quot;)]
    Task&lt;IApiResponse&lt;UserList&gt;&gt; GetUsersAsync();
}

...

// Then we can handle the IApizrResponse&lt;T&gt; response comming from Apizr
var response = await _reqResManager.ExecuteAsync(api =&gt; api.GetUsersAsync());

// Log potential unhandled exceptions and maybe inform the user about it
if(!response.IsSuccess &amp;&amp; !response.Exception.Handled)
{
   _logger.LogError(response.Exception);
    Alert.Show(&quot;Error&quot;, response.Exception.Message);
}

// Use the data, no matter the source
if(response.Result?.Data?.Any() == true)
{
    Users = new ObservableCollection&lt;User&gt;(response.Result.Data);

    // Inform the user that data comes from cache if so
    if(response.DataSource == ApizrResponseDataSource.Cache)
        Toast.Show(&quot;Data comes from cache&quot;);
}
</code></pre>
<h3 id="using-funcapizrexception-bool">Using <code>Func&lt;ApizrException, bool&gt;</code></h3>
<p>Instead of trycatching all the things, you may want to provide an exception handler, thanks to <code>WithExCatching</code> builder option, available at both register and request time.</p>
<p>You can set it thanks to this option:</p>
<pre><code class="lang-csharp">// direct configuration
options =&gt; options.WithExCatching(OnException)
</code></pre>
<div class="tabGroup" id="tabgroup_1">
<ul role="tablist">
<li role="presentation">
<a href="#tabpanel_1_tabid-registering" role="tab" aria-controls="tabpanel_1_tabid-registering" data-tab="tabid-registering" tabindex="0" aria-selected="true">Registering</a>
</li>
<li role="presentation">
<a href="#tabpanel_1_tabid-requesting" role="tab" aria-controls="tabpanel_1_tabid-requesting" data-tab="tabid-requesting" tabindex="-1">Requesting</a>
</li>
</ul>
<section id="tabpanel_1_tabid-registering" role="tabpanel" data-tab="tabid-registering">

<p>Configuring an exception handler at register time allows you to get some Global Exception Handling concepts right in place.</p>
<p><code>WithExCatching</code> builder option is available with or without using registry.
It means that you can share your exception handler globally by setting it at registry level and/or set some specific one at api level.</p>
<p>Here is a quite simple scenario:</p>
<pre><code class="lang-csharp">var reqResUserManager = ApizrBuilder.Current.CreateManagerFor&lt;IReqResUserService&gt;(options =&gt; options
                    .WithExCatching(OnException));

private bool OnException(ApizrException ex)
{
    // this is a global exception handler 
    if (ex.InnerException is IOException)
	{
		// handle no network exception globally for example
        Alert.Show(&quot;No network&quot;, &quot;Please check your connection and try again&quot;);

        // Tell other exception handlers that we handled it yet
		return true;
	}

    return false;
}
</code></pre>
<p>And here is a more complexe scenario:</p>
<pre><code class="lang-csharp">var apizrRegistry = ApizrBuilder.Current.CreateRegistry(
    registry =&gt; registry
        .AddManagerFor&lt;IHttpBinService&gt;()
        .AddCrudManagerFor&lt;User, int, PagedResult&lt;User&gt;, IDictionary&lt;string, object&gt;&gt;(),
        .AddGroup(
            group =&gt; group
                .AddManagerFor&lt;IReqResResourceService&gt;()
                .AddManagerFor&lt;IReqResUserService&gt;(
                    // IReqResUserService dedicated exception handler
                    options =&gt; options.WithExCatching(OnReqResUserException, strategy: ApizrDuplicateStrategy.Add)),

            // Group exception handler common to IReqResUserService &amp; IReqResResourceService apis
            options =&gt; options.WithExCatching(OnGroupException, strategy: ApizrDuplicateStrategy.Add))

    // Global exception handler common to all apis
    options =&gt; options.WithExCatching(OnGlobalException, strategy: ApizrDuplicateStrategy.Add));

private bool OnGlobalException(ApizrException ex)
{
    // this is a global exception handler 
    // called back in case of exception thrown 
    // while requesting with any managed api from the registry
    if (ex.InnerException is IOException)
	{
		// handle no network exception globally for example
        Alert.Show(&quot;No network&quot;, &quot;Please check your connection and try again&quot;);

        // Tell other exception handlers that we handled it yet
		return true;
	}

    return false;
}

private bool OnGroupException(ApizrException ex)
{
    // this is a group exception handler 
    // called back in case of exception thrown 
    // while requesting with any managed api from the group
    if(!ex.Handled) // Not yet handled ?
    {
		// handle it here at group level
        ...
        return true;
	}

    return false;
}

private bool OnReqResUserException(ApizrException ex)
{
    // this is a dedicated exception handler 
    // called back in case of exception thrown 
    // while requesting with IReqResUserService managed api
    if(!ex.Handled) // Not yet handled ?
    {
		// handle it here at api level
        ...
        return true;
	}

    return false;
}
</code></pre>
<p>Here, as I registered callbacks with <code>Add</code> strategy, I'm telling Apizr to:</p>
<ul>
<li>Call back <code>OnGlobalException</code> then <code>OnGroupException</code> then <code>OnReqResUserException</code> in case of any exception thrown while requesting with <code>IReqResUserService</code> api</li>
<li>Call back <code>OnGlobalException</code> then <code>OnGroupException</code> in case of any exception thrown while requesting with <code>IReqResResourceService</code> api</li>
<li>Call back only <code>OnGlobalException</code> in case of any exception thrown while requesting with <code>IHttpBinService</code> api or <code>User</code> CRUD api</li>
</ul>
<p>Feel free to configure your exception handlers at the level of your choice, depending on your needs.
You definitly can mix it all with request option exception handling.</p>
<p>As I leaved the <code>letThrowOnHandledException</code> parameter to its default <code>true</code> value, Apizr will throw back the exception in the end to let you catch it for final specific handling.
But you definitly can tell Apizr to not throw the final exception if yet handled, by setting <code>letThrowOnHandledException</code> parameter to <code>false</code> and then dealing with result default value.</p>
<p>Note that you can mix it too with previous IApizrResponse handling.</p>
</section>
<section id="tabpanel_1_tabid-requesting" role="tabpanel" data-tab="tabid-requesting" aria-hidden="true" hidden="hidden">

<p>Configuring an exception handler at request time allows you to set it at the very end, just before sending the request, like trycatching does.</p>
<pre><code class="lang-csharp">public ObservableCollection&lt;User&gt; Users { get; set; }
...
var reqResManager = apizrRegistry.GetManagerFor&lt;IReqResUserService&gt;();

try
{
	var users = await reqResManager.ExecuteAsync((options, api) =&gt; api.GetUsersAsync(options), 
		options =&gt; options.WithExCatching(OnGetUsersException, strategy: ApizrDuplicateStrategy.Add));

    Users = new ObservableCollection&lt;User&gt;(users);
}
catch (ApizrException&lt;ApiResult&lt;User&gt;&gt; ex)
{
	// handle it here at request level
	if(!ex.Handled) // Not yet handled ?
    {
		// handle it here at result level
        ...
	}
}
...
private bool OnGetUsersException(ApizrException&lt;ApiResult&lt;User&gt;&gt; ex)
{
    // this is a method dedicated exception handler 
    // called back in case of exception thrown 
    // while requesting with a specific managed api's request
    if(!ex.Handled) // Not yet handled ?
    {
		// handle it here at request level
        ...
        return true;
	}

    return false;
}
</code></pre>
<p>Here, as I set the callback with <code>Add</code> strategy, I'm telling Apizr to:</p>
<ul>
<li>Call back any other registered exception handlers (see Registering tab)</li>
<li>Then call back <code>OnGetUsersException</code> (e.g. to display a dedicated message or something)</li>
<li>Then throw back the final exception to catch it for specific handling</li>
</ul>
<p>As I leaved the <code>letThrowOnHandledException</code> parameter to its default <code>true</code> value, Apizr will throw back the exception in the end to let you catch it for final specific handling.
But you definitly can tell Apizr to not throw the final exception if yet handled, by setting <code>letThrowOnHandledException</code> parameter to <code>false</code> and then dealing with result default value.</p>
<p>Note that you can mix it too with previous IApizrResponse handling.</p>
</section>
</div>

<p>You may notice that:</p>
<ul>
<li><code>strategy</code> parameter let you adjust the behavior in case of mixing (default: <code>Replace</code>):
<ul>
<li><code>Ignore</code>: if there's another handler yet configured, ignore this one</li>
<li><code>Add</code>: add/queue this handler, no matter of yet configured ones</li>
<li><code>Replace</code>: replace all yet configured handlers by this one</li>
<li><code>Merge</code>: add/queue this handler, no matter of yet configured ones</li>
</ul>
</li>
<li><code>letThrowOnHandledException</code> parameter tells Apizr to throw the actual exception even if it's handled (default: <code>true</code>)</li>
</ul>
<h3 id="using-optionalasync">Using <code>Optional.Async</code></h3>
<p>Here is how we could handle exceptions using Optional.Async:</p>
<pre><code class="lang-csharp">var optionalUserList = await _reqResOptionalMediator.SendFor(api =&gt; api.GetUsersAsync());

optionalPagedResult.Match(userList =&gt;
{
    if (userList.Data != null &amp;&amp; userList.Data.Any())
        Users = new ObservableCollection&lt;User&gt;(userList.Data);
}, e =&gt;
{
    var message = e.InnerException is IOException ? &quot;No network&quot; : (e.Message ?? &quot;Error&quot;);
    UserDialogs.Instance.Toast(new ToastConfig(message) { BackgroundColor = Color.Red, MessageTextColor = Color.White });

    if (e.CachedResult?.Data != null &amp;&amp; e.CachedResult.Data.Any())
        Users = new ObservableCollection&lt;User&gt;(e.CachedResult.Data);
});
</code></pre>
<p>Optional is pretty cool when trying to handle nullables and exceptions, but what if we still want to write it shorter to get our request done and managed with as less code as possible.
Even if we use the typed optional mediator or typed crud optional mediator to get things shorter, we still have to deal with the result matching boilerplate.
Fortunately, Apizr provides some dedicated extensions to help getting things as short as we can with exceptions handled.</p>
<h4 id="with-onresultasync">With <code>OnResultAsync</code></h4>
<p><code>OnResultAsync</code> ask you to provide one of these parameters:</p>
<ul>
<li><code>Action&lt;TResult&gt; onResult</code>: this action will be invoked just before throwing any exception that might have occurred during request execution
<pre><code class="lang-csharp">await _reqResOptionalMediator.SendFor(api =&gt; api.GetUsersAsync())
    .OnResultAsync(userList =&gt; 
    { 
        users = userList?.Data; 
    });
</code></pre>
</li>
<li><code>Func&lt;TResult, ApizrException&lt;TResult&gt;, bool&gt; onResult</code>: this function will be invoked with the returned result and potential occurred exception
<pre><code class="lang-csharp">await _reqResOptionalMediator.SendFor(api =&gt; api.GetUsersAsync())
    .OnResultAsync((userList, exception) =&gt; 
    { 
        users = userList?.Data; 

        if(exception != null)
            throw exception;

        return true;
    });
</code></pre>
</li>
<li><code>Func&lt;TResult, ApizrException&lt;TResult&gt;, Task&lt;bool&gt;&gt; onResult</code>: this function will be invoked async with the returned result and potential occurred exception
<pre><code class="lang-csharp">var success = await _reqResOptionalMediator.SendFor(api =&gt; api.GetUsersAsync())
    .OnResultAsync((userList, exception) =&gt; 
    { 
        users = userList?.Data; 

        return exception != null;
    });
</code></pre>
</li>
</ul>
<p>All give you a result returned from fetch if succeed, or cache if failed (if configured). The main goal here is to set any binded property with the returned result (fetched or cached), no matter of exceptions. Then the Action will let the exception throw, where the Func will let you decide to throw manually or return a success boolean flag.
Of course, remember to catch throwing exceptions.</p>
<h4 id="with-catchasync">With <code>CatchAsync</code></h4>
<p><code>CatchAsync</code> let you provide these parameters:</p>
<ul>
<li><code>Action&lt;Exception&gt; onException</code>: this action will be invoked just before returning the result from cache if fetch failed. Useful to inform the user of the api call failure and that data comes from cache.</li>
<li><code>letThrowOnExceptionWithEmptyCache</code>: True to let it throw the inner exception in case of empty cache, False to handle it with onException action and return empty cache result (default: False)</li>
</ul>
<p>This one returns result from fetch or cache (if configured), no matter of potential exception handled on the other side by an action callback</p>
<pre><code class="lang-csharp">var userList = await _reqResOptionalMediator.SendFor(api =&gt; api.GetUsersAsync())
    .CatchAsync(AsyncErrorHandler.HandleException, true);
</code></pre>
<p>Here we ask the api to get users and if it fails:</p>
<ul>
<li>There’s some cached data?
<ul>
<li><a href="https://github.com/Fody/AsyncErrorHandler">AsyncErrorHandler</a> will handle the exception like to inform the user that call just failed</li>
<li>Apizr will return the previous result from cache</li>
</ul>
</li>
<li>There’s no cached data yet!
<ul>
<li>letThrowOnExceptionWithEmptyCache is True? (which is the case here)
<ul>
<li>Apizr will throw the inner exception that will be catched further by AsyncErrorHander (this is its normal behavior)</li>
</ul>
</li>
<li>letThrowOnExceptionWithEmptyCache is False! (default)
<ul>
<li>Apizr will return the empty cache data (null) which has to be handled then</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>One line of code to get all the thing done safely and shorter than ever!</p>

</article>

        <div class="contribution d-print-none">
          <a href="https://github.com/Respawnsive/Apizr/blob/dev/Apizr/Docs/Apizr.Docs/articles/config_exceptions.md/#L1" class="edit-link">Edit this page</a>
        </div>

        <div class="next-article d-print-none border-top" id="nextArticle"></div>

      </div>

      <div class="affix">
        <nav id="affix"></nav>
      </div>
    </main>

    <div class="container-xxl search-results" id="search-results"></div>

    <footer class="border-top text-secondary">
      <div class="container-xxl">
        <div class="flex-fill">
          Supported by Respawnsive
        </div>
      </div>
    </footer>
  </body>
</html>
